specs/rollup-node-p2p.md

```diff
@@ -0,0 +1,258 @@
+enforced through a [`StrictNoSign` Signature Policy][signature-policy]
+
+This provides greater privacy, and allows sequencers (consensus identity) to maintain
+multiple network identities for redundancy.
+
+##### Message compression and limits
+
+The application contents are compressed with [snappy][snappy] single-block-compression
+(as opposed to frame-compression), and constrained to 10 MiB.
+
+##### Message ID computation
+
+[Same as L1][l1-message-id], with recognition of compression:
+
+- If `message.data` has a valid snappy decompression, set `message-id` to the first 20 bytes of the `SHA256` hash of
+  the concatenation of `MESSAGE_DOMAIN_VALID_SNAPPY` with the snappy decompressed message data,
+  i.e. `SHA256(MESSAGE_DOMAIN_VALID_SNAPPY + snappy_decompress(message.data))[:20]`.
+- Otherwise, set `message-id` to the first 20 bytes of the `SHA256` hash of
+  the concatenation of `MESSAGE_DOMAIN_INVALID_SNAPPY` with the raw message data,
+  i.e. `SHA256(MESSAGE_DOMAIN_INVALID_SNAPPY + message.data)[:20]`.
+
+#### Heartbeat and parameters
+
+GossipSub [parameters][gossip-parameters]:
+
+- `D` (topic stable mesh target count): 8
+- `D_low` (topic stable mesh low watermark): 6
+- `D_high` (topic stable mesh high watermark): 12
+- `D_lazy` (gossip target): 6
+- `heartbeat_interval` (frequency of heartbeat, seconds): 0.5
+- `fanout_ttl` (ttl for fanout maps for topics we are not subscribed to but have published to, seconds): 24
+- `mcache_len` (number of windows to retain full messages in cache for `IWANT` responses): 12
+- `mcache_gossip` (number of windows to gossip about): 3
+- `seen_ttl` (number of heartbeat intervals to retain message IDs): 40
+
+Notable differences from L1 consensus (Eth2):
+
+- `seen_ttl` does not need to cover a full L1 epoch (6.4 minutes), but rather just a small window covering latest blocks
+- `fanout_ttl`: adjusted to lower than `seen_ttl`
+- `mcache_len`: a larger number of heartbeats can be retained since the gossip is much less noisy.
+- `heartbeat_interval`: faster interval to reduce latency, bandwidth should still be reasonable since
+  there are far fewer messages to gossip about each interval than on L1 which uses an interval of 0.7 seconds.
+
+#### Topic configuration
+
+Topics have string identifiers and are communicated with messages and subscriptions.
+`/optimism/chain_id/hardfork_version/Name`
+
+- `chain_id`: replace with decimal representation of chain ID
+- `hardfork_version`: replace with decimal representation of hardfork
+- `Name`: topic application-name
+
+Note that the topic encoding depends on the topic, unlike L1,
+since there are less topics, and all are snappy-compressed.
+
+## Gossip Topics
+
+### `blocks`
+
+The primary topic of the L2, to distribute blocks to other nodes faster than proxying through L1 would.
+
+#### Block encoding
+
+TODO: encode execution payload (SSZ or RLP), with sequencer identifier and signature.
```

Comment:
Should the p2p format be any different than the L1 block format RLP encoded with the signature in the extradata?

Reply:
Yes, we don't intend to use `extradata`, at all. These signatures need to only be available temporarily, they are only required against DoS/spam protection during gossip, not during sync. And yes, we can have a different key for gossip than for batch-submission.

So instead of `extradata` like Clique, we just encode the block (in whatever format, RLP is fine), and wrap it with an envelope container type that adds a signature.

Also note that we don't use the total-difficulty like clique to figure out the head of the chain. We just take the first block received from the sequencer (while there's only one sequencer this is simple, at worst there's equivocation with multiple different blocks, but we'll check against L1 anyway) and try to apply it to the engine as unsafe-block.